### PR TITLE
feat: Named accessors with multiple variants

### LINKED ISSUE #935: Generate accessors for structurally compatible multi-variant custom types
As @CrowdHailer put it
> If you have
> ```rust
> type Foo {
>   A(id: Int)
>   B(id: Int)
>}
>```
>
> Should you be able to call?
> ```rust
> let foo = A(3)
> foo.id
>```

### PR DESCRIPTION
**Pull request by HarryET**
Fixes #935 

## Tests Needed
- [x] Accessor cannot be used if type is different
- [x] Accessor cannot be used if label is different
- [x] Accessor can be used in position other than 1st field
- [x] Accessor can be used in position other than 1st field when 1st field of each variant is different
- [x] Test showing Erlang code generator
- [x] Test showing JavaScript code generation

### CODE REVIEW

**Review by lpil — COMMENTED** [review]
Looking great!

We need a few more tests to cover other edge cases:

- Accessor cannot be used if type is different
- Accessor cannot be used if index is different
- Accessor can be used in position other than 1st field
- Accessor can be used in position other than 1st field when 1st field of each variant is different
- Test showing Erlang code generator
- Test showing JavaScript code generation

File: compiler-core/src/type_.rs [hunk 1]
```diff
@@ -1579,18 +1579,34 @@ fn make_type_vars(
...
+
+    let (first_constructor_arg, constructor_args) = constructor_args_data
+        .split_first()
+        .expect("No constructs with args");
+    let args = first_constructor_arg
+        .iter()
+        .filter(|data_type| {
+            constructor_args.iter().all(|set| {
+                set.iter()
+                    .map(|item| (item.label.clone(), item.type_.clone()))
```
Comment (lpil) [inline compiler-core/src/type_.rs:1605]: I think we can do this without cloning the label

Comment (HarryET) [reply compiler-core/src/type_.rs:1605]: Switched to `.as_ref()`

File: compiler-core/src/type_.rs [hunk 2]
```diff
@@ -1579,18 +1579,34 @@ fn make_type_vars(
...
+        .map(|constructor| constructor.arguments.as_slice())
+        .collect_vec();
+
+    let (first_constructor_arg, constructor_args) = constructor_args_data
+        .split_first()
+        .expect("No constructs with args");
+    let args = first_constructor_arg
+        .iter()
+        .filter(|data_type| {
+            constructor_args.iter().all(|set| {
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/type_.rs:1603]: Renamed

File: compiler-core/src/type_.rs [hunk 3]
```diff
@@ -1579,18 +1579,39 @@ fn make_type_vars(
         .try_collect()
 }
 
-fn custom_type_accessors<A>(
+#[allow(unused_qualifications)]
```
Comment (lpil) [inline compiler-core/src/type_.rs:1582]: We should never disable the linter, best fix this.

Comment (HarryET) [reply compiler-core/src/type_.rs:1582]: Without that I get this:
```
warning: unnecessary qualification                                                                                                                                                                                                     
    --> compiler-core\src\type_.rs:1582:37
     |
1582 | fn custom_type_accessors<A: Clone + std::cmp::PartialEq>(
     |                                     ^^^^^^^^^^^^^^^^^^^
     |
note: the lint level is defined here
    --> compiler-core\src\lib.rs:33:5
     |
33   |     unused_qualifications,
     |     ^^^^^^^^^^^^^^^^^^^^^
   ```
  I then remove it to get this on the next build:
  ```
PS B:\Development\gleam-lang> cargo build
   Compiling gleam-core v0.21.0-rc1 (B:\Development\gleam-lang\compiler-core)
error[E0369]: binary operation `==` cannot be applied to type `&ast::RecordConstructorArg<A>`                                                                                                                                          
    --> compiler-core\src\type_.rs:1617:30
     |
1617 |                         item == &clear_data_type
     |                         ---- ^^ ---------------- &ast::RecordConstructorArg<A>
     |                         |
     |                         &ast::RecordConstructorArg<A>
     |
help: consider further restricting this bound
     |
1582 | fn custom_type_accessors<A: Clone + std::cmp::PartialEq>(
     |                                   +++++++++++++++++++++

For more information about this error, try `rustc --explain E0369`.                                                                                                                                                                    
error: could not compile `gleam-core` due to previous error
```

File: compiler-core/src/ast.rs [hunk 4]
```diff
@@ -282,6 +282,41 @@ impl TypeAst {
             | TypeAst::Constructor { location, .. } => *location,
         }
     }
+
+    pub fn clear_location(&self) -> TypeAst {
```
Comment (lpil) [inline compiler-core/src/ast.rs:286]: This is expensive, we shouldn't be making a new copy of the TypeAst, instead we should have a comparison function

File: compiler-core/src/type_.rs (hunk 3 above)
Comment (lpil) [reply compiler-core/src/type_.rs:1582]: It's asking you to remove `std::cmp::`, that's the qualification. If you `cmd+.` on it in vscode it'll be able to fix it for you

Comment (HarryET) [reply compiler-core/src/type_.rs:1582]: Oh sorry, will try now. I'm currently using CLion and it isn't helping with that 😅

File: compiler-core/src/ast.rs (hunk 4 above)
Comment (HarryET) [reply compiler-core/src/ast.rs]: Fixed to use comparison functions

File: compiler-core/src/ast.rs [hunk 5]
```diff
@@ -282,6 +289,47 @@ impl TypeAst {
...
         }
     }
+
+    pub fn compare_without_location(&self, other: &TypeAst) -> bool {
+        match self {
+            TypeAst::Constructor {
+                module,
+                name,
+                arguments,
+                ..
```
Comment (lpil) [inline compiler-core/src/ast.rs:299]: Can we replace the `..` with explicit `thingy: _`? I want to be alerted to future fields being added.

File: compiler-core/src/erlang/tests/records.rs [hunk 6]
```diff
@@ -93,6 +93,34 @@ pub fn get_name(person: Person) { person.name }
...
+    // In positions other than the 1st field
+    assert_erl!(
+        "
+pub type Person {
+    Teacher(name: String, age: Int, title: String)
+    Student(name: String, age: Int)
+}
+pub fn get_name(person: Person) { person.name }
+pub fn get_age(person: Person) { person.age }"
+    );
```
Comment (lpil) [inline compiler-core/src/erlang/tests/records.rs:121]: Could we also have a test like this where the first is different? Thanks

```gleam
pub type Person {
    Teacher(name: Nil, age: Int)
    Student(name: String, age: Int)
}
pub fn get_age(person: Person) { person.age }
```

**Review by lpil — COMMENTED** [review]
Fantastic work, really loving all the tests. I've left 2 small comments

File: compiler-core/src/erlang/tests/records.rs (hunk 6 above)
Comment (HarryET) [reply compiler-core/src/erlang/tests/records.rs:121]: For both erlang and js?

File: compiler-core/src/type_.rs [hunk 7]
```diff
@@ -1579,28 +1579,50 @@ fn make_type_vars(
...
+                    .collect_vec()
+                    .is_empty()
+            })
+        })
+        .collect::<Vec<_>>();
 
     let mut fields = HashMap::with_capacity(args.len());
     hydrator.disallow_new_type_variables();
-    for (index, RecordConstructorArg { label, ast, .. }) in args.iter().enumerate() {
+    for (_, (index, RecordConstructorArg { label, ast, .. })) in args.iter().enumerate() {
```
Comment (lpil) [inline compiler-core/src/type_.rs:1619]:
```suggestion
    for (index, RecordConstructorArg { label, ast, .. }) in args.iter() {
```

**Review by lpil — APPROVED** [review]
Brilliant!
